test/Table_Tests/src/Common_Table_Operations/Union_Spec.enso

```diff
@@ -0,0 +1,278 @@
+            t6.at "mixed" . value_type . should_equal Value_Type.Mixed
+
+        Test.specify "if no common type can be found, should report error and drop the problematic column" <|
+            t1 = table_builder [["A", [1, 2, 3]], ["B", ["a", "b", "c"]], ["C", [True, False, Nothing]]]
+            t2 = table_builder [["C", ["x", "Y", "Z"]], ["A", [4, 5, 6]], ["B", [1, 2, 3]]]
+
+            action = t1.union t2 on_problems=_
+            tester table =
+                expect_column_names ["A"] table
+                table.at "A" . to_vector . should_equal [1, 2, 3, 4, 5, 6]
+            problems = [No_Common_Type.Error "B", No_Common_Type.Error "C"]
+            Problems.test_problem_handling action problems tester
+
+        Test.specify "if type widening is not allowed, should use the type from first table that contained the given column" <|
+            t1 = table_builder [["A", [1, 2, 3]], ["B", [1.5, 2.5, 3.5]]]
+            t2 = table_builder [["A", [4, 5, 6]], ["B", [1, 2, 3]]]
+
+            t1.at "B" . value_type . should_equal Value_Type.Float
+            t2.at "B" . value_type . should_equal Value_Type.Integer
+
+            t3 = t1.union t2 allow_type_widening=False
+            within_table t3 <|
+                Problems.assume_no_problems t3
+                expect_column_names ["A", "B"] t3
+                t3.at "A" . to_vector . should_equal [1, 2, 3, 4, 5, 6]
+                t3.at "B" . to_vector . should_equal [1.5, 2.5, 3.5, 1, 2, 3]
+                t3.at "B" . value_type . should_equal Value_Type.Float
+
+        Test.specify "if type widening is not allowed and types do not match, should report error and drop the problematic column" <|
+            t1 = table_builder [["A", [1, 2, 3]], ["B", [1, 2, 3]], ["C", [True, False, Nothing]], ["D", [10, 20, 30]]]
+            t2 = table_builder [["A", [4, 5, 6]], ["B", [1.5, 2.5, 3.5]], ["C", [1, 2, 3]], ["D", [True, True, True]]]
+
+            t1.at "B" . value_type . should_equal Value_Type.Float
+            t1.at "C" . value_type . should_equal Value_Type.Boolean
+            t1.at "D" . value_type . should_equal Value_Type.Integer
+
+            t2.at "B" . value_type . should_equal Value_Type.Integer
+            t2.at "C" . value_type . should_equal Value_Type.Integer
+            t2.at "D" . value_type . should_equal Value_Type.Boolean
+
+            action = t1.union t2 allow_type_widening=False on_problems=_
+            tester table =
+                expect_column_names ["A"] table
+                table.at "A" . to_vector . should_equal [4, 5, 6, 1, 2, 3]
+            problems = [Column_Type_Mismatch.Error "B" Value_Type.Integer Value_Type.Float, Column_Type_Mismatch.Error "C" Value_Type.Boolean Value_Type.Integer, Column_Type_Mismatch.Error "D" Value_Type.Integer Value_Type.Boolean]
+            Problems.test_problem_handling action problems tester
```

Comment:
I'm not sure about the rules when type widening is disabled.

In the spec I wrote there, that with type widening off, no widening is performed, but we _do_ allow widening an integer column to a floating point column. I'm not certain anymore this is good - once I implemented this I realised this same logic applies to Booleans - in that case if I start with Integer column and append a Boolean column to it, I can also cast Booleans to 0 and 1 and proceed. But that is definitely widening - so at this point `allow_type_widening` would not really be controlling widening but whether the first type decides about result type or if all column types are unified.

I'm not sure why the `int->float` widening should be somehow special and skip `allow_type_widening=False` when `bool->int` does not do that.

So to be consistent, I think in `allow_type_widening=False` mode, no widening _at all_ should be allowed - even if an integer column is appended to a floating point one. Instead, in this mode I suggest we just check `value_type` of all columns and verify that all of them are _exactly_ equal. If there is any inequality, we'd throw `Column_Type_Mismatch`. I think this way this will be much more consistent and more logical - thus easier to understand by the user.

@jdunkerley do you think this change of semantics is acceptable?